websocket_server/Stores/RoomStateStore.js

```diff
@@ -0,0 +1,359 @@
+	}
+
+	#expiresAt(ttlMs) {
+		if (!ttlMs) return null
+		return Date.now() + ttlMs
+	}
+
+	#ttlSeconds(ttlMs) {
+		if (!ttlMs) return null
+		return Math.max(1, Math.ceil(ttlMs / 1000))
+	}
+
+	#isExpired(expiresAt) {
+		return expiresAt !== null && Date.now() >= expiresAt
+	}
+
+	#stringify(value) {
+		return JSON.stringify(value)
+	}
+
+	#parse(value) {
+		try {
+			return JSON.parse(value)
+		} catch (e) {
+			return null
+		}
+	}
+
+	#matchesPattern(key, pattern) {
+		if (!pattern || pattern === '*') {
+			return true
+		}
+		const escaped = pattern.replace(/[-/\\^$+?.()|[\]{}]/g, '\\$&')
+		const regex = new RegExp(`^${escaped.replace(/\*/g, '.*')}$`)
+		return regex.test(key)
+	}
+
+	async listValueKeys(matchPattern = '*') {
+		if (this.shouldUseRedis()) {
+			const keys = []
+			const fullMatch = this.#fullKey(matchPattern)
+			for await (const key of this.redisClient.scanIterator({ MATCH: fullMatch })) {
+				keys.push(key.slice(this.prefix.length))
+			}
+			return keys
+		}
+
+		return Array.from(this.memoryStore.keys())
+			.map((key) => key.slice(this.prefix.length))
+			.filter((key) => this.#matchesPattern(key, matchPattern))
+	}
+
+	async listHashKeys(matchPattern = '*') {
+		if (this.shouldUseRedis()) {
+			const keys = []
+			const fullMatch = this.#fullKey(matchPattern)
+			for await (const key of this.redisClient.scanIterator({ MATCH: fullMatch })) {
+				keys.push(key.slice(this.prefix.length))
+			}
+			return keys
+		}
+
+		return Array.from(this.memoryHashes.keys())
+			.map((key) => key.slice(this.prefix.length))
+			.filter((key) => this.#matchesPattern(key, matchPattern))
+	}
+
+	async getValue(key, { isStale } = {}) {
+		const fullKey = this.#fullKey(key)
+
+		if (this.shouldUseRedis()) {
+			try {
+				const raw = await this.redisClient.get(fullKey)
+				if (!raw) return null
+				const value = this.#parse(raw)
+				if (isStale && value && await isStale(value)) {
+					await this.redisClient.del(fullKey)
+					return null
+				}
+				return value
+			} catch (error) {
+				if (this.#isClientClosedError(error)) {
+					return null
+				}
+				throw error
+			}
+		}
+
+		const entry = this.memoryStore.get(fullKey)
+		if (!entry || this.#isExpired(entry.expiresAt)) {
+			this.memoryStore.delete(fullKey)
+			return null
+		}
+		if (isStale && entry.value && await isStale(entry.value)) {
+			this.memoryStore.delete(fullKey)
+			return null
+		}
+		return entry.value
+	}
+
+	async setValue(key, value, { ttlMs } = {}) {
+		const fullKey = this.#fullKey(key)
+		const ttlToUse = ttlMs || this.defaultTtlMs
+
+		if (this.shouldUseRedis()) {
+			try {
+				const serialized = this.#stringify(value)
+				const ttlSeconds = this.#ttlSeconds(ttlToUse)
+				if (ttlSeconds) {
+					await this.redisClient.set(fullKey, serialized, { EX: ttlSeconds })
+				} else {
+					await this.redisClient.set(fullKey, serialized)
+				}
+				return
+			} catch (error) {
+				if (this.#isClientClosedError(error)) {
+					return
```

Comment:
This may hide a critical error, @benjaminfrueh suggested that we may still throw here instead